tao_compiler/mlir/disc/transforms/disc_side_effect_loop_invariant_code_motion.cc

```diff
@@ -0,0 +1,119 @@
+// Copyright 2022 The BladeDISC Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This pass makes `loop invariant code motion` optimization on operators with
+// side effect.
+
+#include "mlir-hlo/Dialect/lhlo/IR/lhlo_ops.h"
+#include "mlir/Dialect/SCF/SCF.h"                        // from @llvm-project
+#include "mlir/IR/Location.h"                            // from @llvm-project
+#include "mlir/IR/MLIRContext.h"                         // from @llvm-project
+#include "mlir/IR/Operation.h"                           // from @llvm-project
+#include "mlir/IR/PatternMatch.h"                        // from @llvm-project
+#include "mlir/Interfaces/LoopLikeInterface.h"           // from @llvm-project
+#include "mlir/Pass/Pass.h"                              // from @llvm-project
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+#include "mlir/Transforms/Passes.h"                      // from @llvm-project
+#include "tensorflow/compiler/mlir/disc/disc_util.h"
+#include "transforms/PassDetail.h"
+
+namespace mlir {
+namespace disc_ral {
+
+namespace {
+
+// Move the invariant memref::load op out of the loop.
+template <typename LoopOp>
+struct LoadLoopInvariantCodeMotion : public OpRewritePattern<LoopOp> {
+  explicit LoadLoopInvariantCodeMotion(MLIRContext* context)
+      : OpRewritePattern<LoopOp>(context) {}
+
+  LogicalResult matchAndRewrite(LoopOp op,
+                                PatternRewriter& rewriter) const override {
+    LoopLikeOpInterface loop = dyn_cast<LoopLikeOpInterface>(op.getOperation());
+    if (!loop) {
+      return failure();
+    }
+
+    auto valueKnownNotModifiedInLoop = [&](Value value) {
+      return llvm::all_of(value.getUsers(), [&](Operation* user) {
+        // Check if `user` is outside of the loop.
+        if (loop.getLoopBody().findAncestorOpInRegion(*user) == nullptr) {
+          return true;
+        }
+
+        // Skip if there are memref alias operators.
+        if (disc_ral::IsMemRefAliasOp(user)) {
```

Comment:
nit: to my understanding, it's still possible to analyses alias of users recursively to move more op outside?

Reply:
Yes, you are right! I do not do that because most alias ops are already moved out by the non-side-effect-LICM pass and I do not observe alias ops left for softmax and layernorm operators. But surely there are potentially some more opportunities by analyzing alias chains.